### PR TITLE
[Feature-3189][alert,spi,dao,plugin-api] base code of dolphinscheduler spi and alert plugin implement

### PR DESCRIPTION
**Pull request by gaojun2048**
## What is the purpose of the pull request

https://github.com/apache/incubator-dolphinscheduler/issues/3189

This pull request contain the base code of dolphinscheduler spi . And designed the spi of the alert module and implemented the email alert plugin based on the alert spi.

## Brief change log

* Remove dolphinscheduler-plugin-api module

* Add dolphinscheduler-spi and dolphinscheduler-alert-plugin module

* And designed the spi of the alert module and implemented the email alert plugin based on the alert spi.

## notice

* Some class need update but because I think it shoud be update in other sub task of https://github.com/apache/incubator-dolphinscheduler/issues/3049 . So I add some TODO flag in the class, I think this can help the contribute who work for the issue.

### CODE REVIEW

File: sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+CREATE TABLE `t_ds_plugin_define` (
+  `id` int NOT NULL AUTO_INCREMENT,
+  `plugin_name` varchar(100) NOT NULL COMMENT 'the name of plugin eg: email',
+  `plugin_type` varchar(100) NOT NULL COMMENT 'plugin type . alert=alert plugin, job=job plugin',
+  `plugin_params` text COMMENT 'plugin params',
+  `create_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `update_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `t_ds_plugin_define_UN` (`plugin_name`,`plugin_type`)
+) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8
```
Comment (yangyichao-mango) [inline sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql:29]:
```suggestion
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8;
```

File: sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+CREATE TABLE `t_ds_alert_plugin_instance` (
+  `id` int NOT NULL AUTO_INCREMENT,
+  `plugin_define_id` int NOT NULL,
+  `plugin_instance_params` text COMMENT 'plugin instance params. Also contain the params value which user input in web ui.',
+  `create_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
+  `update_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  `alert_group_id` int DEFAULT NULL,
+  `instance_name` varchar(200) DEFAULT NULL COMMENT 'alert instance name',
+  PRIMARY KEY (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8
```
Comment (yangyichao-mango) [inline sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql:41]:
```suggestion
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

File: sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql (hunk 1 above)
Comment (gaojun2048) [reply sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql]: ok, I have corrected.

File: sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql (hunk 2 above)
Comment (gaojun2048) [reply sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql]: I have corrected

**Comment by gaojun2048** [comment]
> Your pr caused the workflow to be uneditable, please refer to the screenshot below to fix
> image

Yes, I merged this code from dev .

**Pull request closed without merging**